src/app/ConcreteCommandPath.h

```diff
@@ -0,0 +1,47 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/util/basic-types.h>
+#include <lib/core/Optional.h>
+
+namespace chip {
+namespace app {
+
+/**
+ * A representation of a concrete invoke path, with one additional piece of data
+ * added: the command id of the spec-defined response command, if there is one.
+ */
+struct ConcreteCommandPath
+{
+    ConcreteCommandPath(EndpointId aEndpointId, ClusterId aClusterId, CommandId aCommandId) :
+        mEndpointId(aEndpointId), mClusterId(aClusterId), mCommandId(aCommandId)
+    {}
+
+    ConcreteCommandPath(EndpointId aEndpointId, ClusterId aClusterId, CommandId aCommandId, CommandId aResponseId) :
+        mEndpointId(aEndpointId), mClusterId(aClusterId), mCommandId(aCommandId), mResponseCommand(aResponseId)
+    {}
+
+    const EndpointId mEndpointId = 0;
+    const ClusterId mClusterId   = 0;
+    const CommandId mCommandId   = 0;
+    const Optional<CommandId> mResponseCommand;
```

Comment:
This representation of combining both the request and response IDs in a single 'command path' encapsulation feels a little odd to me, since it doesn't quite match up with the spec.

I'd rather `mCommandId` be modal in its interpretation and be used in both the receive path (to signal an ID in the request) as well as on the transmit side (to signal an ID in the response).

Reply:
Yep, per discussion on Slack and some followup work I'll remove this member.